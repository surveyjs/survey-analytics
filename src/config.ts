/**
 * An interface that describes a visualization item (chart, gauge, etc.).
 *
 * To access `IVisualizerPanelElement` objects, you can use the following properties and methods of `VisualizationPanel`:
 *
 * - [`getElements()`](https://surveyjs.io/dashboard/documentation/api-reference/visualizationpanel#getElements)
 * - [`visibleElements`](https://surveyjs.io/dashboard/documentation/api-reference/visualizationpanel#visibleElements)
 * - [`hiddenElements`](https://surveyjs.io/dashboard/documentation/api-reference/visualizationpanel#hiddenElements)
 */
export interface IVisualizerPanelElement {
  /**
   * The name of a survey question visualized by this item.
   */
  name: string;
  /**
   * The title of a survey question visualized by this item. The visualization item displays the same title.
   */
  displayName: string;
  /**
   * Indicates whether the visualization item is currently visible.
   *
   * If you want to disallow users to hide visualization items, set the [`allowHideQuestions`](https://surveyjs.io/dashboard/documentation/api-reference/ivisualizationpaneloptions#allowHideQuestions) property to `false`.
   */
  isVisible: boolean;
  isPublic: boolean;
  /**
   * Indicates the selected chart type.
   *
   * Possible values:
   *
   * - `"bar"`
   * - `"vbar"`
   * - `"stackedbar"`
   * - `"pie"`
   * - `"doughnut"`
   * - `"gauge"`
   * - `"bullet"`
   */
  chartType?: string;
  /**
   * Indicates the answer order in [bar charts](https://surveyjs.io/dashboard/documentation/chart-types#bar-chart), [histograms](https://surveyjs.io/dashboard/documentation/chart-types#histogram), and [statistics tables](https://surveyjs.io/dashboard/documentation/chart-types#statistics-table).
   *
   * Possible values:
   *
   * - `"default"` - Answers retain the initial sort order.
   * - `"asc"` - Answers are sorted by ascending answer count.
   * - `"desc"` - Answers are sorted by descending answer count.
   */
  answersOrder?: string;
  /**
   * Indicates whether answers with zero count are hidden in [bar charts](https://surveyjs.io/dashboard/documentation/chart-types#bar-chart), [histograms](https://surveyjs.io/dashboard/documentation/chart-types#histogram), and [statistics tables](https://surveyjs.io/dashboard/documentation/chart-types#statistics-table).
   */
  hideEmptyAnswers?: boolean;
  /**
   * Indicates whether the visualization item displays top 5, 10, or 20 answers by answer count. This property equals -1 if the Top N Answers feature is disabled.
   */
  topN?: number;
}

export interface IState {
  locale?: string;
  elements?: IVisualizerPanelElement[];
}

export interface IPermission {
  name: string;
  isPublic: boolean;
}
