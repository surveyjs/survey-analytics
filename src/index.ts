export * from "./localizationManager";

//localization
import "./analytics-localization/farsi";
import "./analytics-localization/french";
import "./analytics-localization/norwegian";
import "./analytics-localization/portuguese";
import "./analytics-localization/russian";
import "./analytics-localization/dutch";
import "./analytics-localization/arabic";
import "./analytics-localization/japanese";

export * from "./dataProvider";

export * from "./visualizerBase";
export * from "./visualizationManager";
export * from "./visualizationPanel";
export * from "./visualizationPanelDynamic";
export * from "./visualizationMatrixDynamic";
export * from "./visualizationMatrixDropdown";
export * from "./alternativeVizualizersWrapper";

export * from "./tables/tabulator";

export * from "./plotly/index";
export * from "./wordcloud/wordcloud";
export * from "./wordcloud/stopwords/index";
export * from "./text";
