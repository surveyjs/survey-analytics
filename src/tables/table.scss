@mixin arrow() {
  display: inline-block;
  background: transparent;
  background-image: url("data:image/svg+xml,%3C%3Fxml version='1.0' encoding='utf-8'%3F%3E%3C!-- Generator: Adobe Illustrator 21.1.0, SVG Export Plug-In . SVG Version: 6.00 Build 0) --%3E%3Csvg version='1.1' id='Layer_1' xmlns='http://www.w3.org/2000/svg' xmlns:xlink='http://www.w3.org/1999/xlink' x='0px' y='0px' viewBox='0 0 10 10' style='enable-background:new 0 0 10 10;' xml:space='preserve'%3E%3Cstyle type='text/css'%3E .st0%7Bfill:%23a0a0a0;%7D%0A%3C/style%3E%3Cpolygon class='st0' points='2,2 0,4 5,9 10,4 8,2 5,5 '/%3E%3C/svg%3E%0A");
  background-repeat: no-repeat, repeat;
  background-size: 9px 100%;
  height: 10px;
  width: 10px;
  white-space: nowrap;
  overflow: hidden;
  text-indent: -9999px;
}

@mixin dropdown() {
  appearance: none;
  -webkit-appearance: none;
  -moz-appearance: none;
  background: transparent;
  background-image: url("data:image/svg+xml,%3C%3Fxml version='1.0' encoding='utf-8'%3F%3E%3C!-- Generator: Adobe Illustrator 21.1.0, SVG Export Plug-In . SVG Version: 6.00 Build 0) --%3E%3Csvg version='1.1' id='Layer_1' xmlns='http://www.w3.org/2000/svg' xmlns:xlink='http://www.w3.org/1999/xlink' x='0px' y='0px' viewBox='0 0 10 10' style='enable-background:new 0 0 10 10;' xml:space='preserve'%3E%3Cstyle type='text/css'%3E .st0%7Bfill:%23404040;%7D%0A%3C/style%3E%3Cpolygon class='st0' points='2,2 0,4 5,9 10,4 8,2 5,5 '/%3E%3C/svg%3E%0A");
  background-repeat: no-repeat, repeat;
  background-position: right 0.7em top 50%, 0 0;
  background-size: 0.57em 100%;
  height: 34px;
  border-radius: 2px;
  border: solid 1px $border-color;
  box-sizing: border-box;
  font-family: inherit;
  font-size: inherit;
  padding: 0.25em 2em 0.25em 0.87em;

  &:focus {
    background-image: url("data:image/svg+xml,%3C%3Fxml version='1.0' encoding='utf-8'%3F%3E%3C!-- Generator: Adobe Illustrator 21.1.0, SVG Export Plug-In . SVG Version: 6.00 Build 0) --%3E%3Csvg version='1.1' id='Layer_1' xmlns='http://www.w3.org/2000/svg' xmlns:xlink='http://www.w3.org/1999/xlink' x='0px' y='0px' viewBox='0 0 10 10' style='enable-background:new 0 0 10 10;' xml:space='preserve'%3E%3Cstyle type='text/css'%3E .st0%7Bfill:%231AB394;%7D%0A%3C/style%3E%3Cpolygon class='st0' points='2,2 0,4 5,9 10,4 8,2 5,5 '/%3E%3C/svg%3E ");
    border-color: $border-color;
    outline: none;
  }
  &::-ms-expand {
    display: none;
  }
}

@mixin input() {
  height: 34px;
  border: none;
  border-bottom: 0.06em solid $border-color;
  box-sizing: border-box;
  font-family: inherit;
  font-size: inherit;
  padding: 0.25em 1.5em 0.25em 0.87em;
  background: transparent;

  &:focus {
    border-color: $main-color;
    outline: none;
  }
}

.sa-table {
  font-size: $root-font-size;
  font-family: $font-family;

  button {
    cursor: pointer;
    background-color: inherit;
    outline: none;
  }

  select {
    @include dropdown();
  }

  input {
    @include input();
  }
}

.sa-table__detail-row.sa-table__detail-row.sa-table__detail-row.sa-table__detail-row {
  background-color: $detail-color;
}

.sa-table__detail-table.sa-table__detail-table {
  width: 100%;
  border-collapse: collapse;
  border-right: 1px solid $border-color;
  border-bottom: none;
}

.sa-table__detail {
  border-top: 1px solid $border-color;
  td {
    box-sizing: border-box;
    padding: 20px;
    border-left: none;
    border-right: none;
    white-space: initial;
  }

  td:first-child {
    width: 267px;
    border-left: none;
    font-weight: bold;
  }

  td:nth-child(2) {
    width: 208px;
  }
}

.sa-table__btn {
  height: 34px;
  width: 154px;
  font-size: $root-font-size;
  font-weight: bold;
  background-color: transparent;
  box-sizing: border-box;
  border: solid 2px;
  border-radius: 17px;
  padding: 0;
  outline: none;
}

a.sa-table__btn {
  display: inline-block;
  text-align: center;
  padding-top: 5px;
}

.sa-table__btn--small {
  width: 99px;
}

.sa-table__btn--gray {
  border-color: $text-gray;
  color: $text-gray;
  &:hover {
    background-color: $dark-gray;
    color: $text-color-inverse;
  }
}

.sa-table__btn--green {
  border-color: $main-color;
  color: $main-color;
  &:hover {
    background-color: $main-color;
    color: $text-color-inverse;
  }
}

.sa-table__svg-button {
  width: 15px;
  height: 15px;
  padding: 0;
  border: none;
  use {
    fill: $svg-color;
  }
  &:hover {
    use {
      fill: $svg-hover-color;
    }
  }
  svg {
    width: 15px;
    height: 15px;
  }
}

.sa-table__action-container {
  margin-bottom: 10px;
  white-space: normal;
  .sa-table__svg-button {
    float: right;
    margin: 0 0 13px 13px;
    pointer-events: all;
  }
  &:after {
    content: "";
    display: table;
    clear: both;
  }
  .sa-table__drag-button {
    float: left;
    cursor: move;
    use {
      fill: $main-color;
    }
    &:hover use {
      fill: $main-color;
    }
  }
}

.sa-table__global-filter.sa-table__global-filter,
.sa-table__filter.sa-table__filter {
  background: transparent;
  background-image: url("data:image/svg+xml;charset=utf8,%3C?xml version='1.0' encoding='utf-8'?%3E%3C!-- Generator: Adobe Illustrator 24.0.1, SVG Export Plug-In . SVG Version: 6.00 Build 0) --%3E%3Csvg version='1.1' id='Layer_1' xmlns='http://www.w3.org/2000/svg' xmlns:xlink='http://www.w3.org/1999/xlink' x='0px' y='0px' viewBox='0 0 16 16' style='enable-background:new 0 0 16 16;' xml:space='preserve'%3E%3Cstyle type='text/css'%3E .st0{fill:%239F9E9E;} %3C/style%3E%3Cpath class='st0' d='M13.56,11.44L11.3,9.18C11.74,8.38,12,7.47,12,6.5C12,3.47,9.53,1,6.5,1S1,3.47,1,6.5C1,9.53,3.47,12,6.5,12 c0.97,0,1.88-0.26,2.68-0.7l2.26,2.26c0.29,0.29,0.66,0.43,1.04,0.44c0.39,0.01,0.78-0.14,1.08-0.44c0.3-0.3,0.44-0.69,0.44-1.08 C13.99,12.1,13.85,11.72,13.56,11.44z M4,6.5C4,5.12,5.12,4,6.5,4S9,5.12,9,6.5S7.88,9,6.5,9S4,7.88,4,6.5z'/%3E%3C/svg%3E");
  background-repeat: no-repeat, repeat;
  background-position: left 0.7em top 50%, 0 0;
  background-size: 1em 100%;
  padding-left: 35px;
  border: solid 1px $border-color;
  pointer-events: all;
}

.sa-table__filter.sa-table__filter {
  width: 100%;
}

.sa-table__show-column.sa-table__show-column {
  background: transparent;
  background-image: url("data:image/svg+xml,%3C%3Fxml version='1.0' encoding='utf-8'%3F%3E%3C!-- Generator: Adobe Illustrator 24.0.1, SVG Export Plug-In . SVG Version: 6.00 Build 0) --%3E%3Csvg version='1.1' id='Layer_1' xmlns='http://www.w3.org/2000/svg' xmlns:xlink='http://www.w3.org/1999/xlink' x='0px' y='0px' viewBox='0 0 16 16' style='enable-background:new 0 0 16 16;' xml:space='preserve'%3E%3Cstyle type='text/css'%3E .st0%7Bfill:rgba(159, 159, 159, 0.5);%7D .st1%7Bfill:rgba(159, 159, 159, 0.5);%7D%0A%3C/style%3E%3Cg%3E%3Cpath class='st1' d='M8,2.03C3.63,2.03,0,8,0,8s3.63,6.03,8,6.03s8-6,8-6S12.37,2.03,8,2.03z M8,12c-2.58,0-5.13-2.81-6-4 c1.07-1.47,3.61-4,6-4c2.58,0,5.13,2.81,6,4C12.93,9.47,10.39,12,8,12z'/%3E%3Ccircle class='st1' cx='8' cy='8' r='3'/%3E%3C/g%3E%3C/svg%3E"),
    url("data:image/svg+xml,%3C%3Fxml version='1.0' encoding='utf-8'%3F%3E%3C!-- Generator: Adobe Illustrator 21.1.0, SVG Export Plug-In . SVG Version: 6.00 Build 0) --%3E%3Csvg version='1.1' id='Layer_1' xmlns='http://www.w3.org/2000/svg' xmlns:xlink='http://www.w3.org/1999/xlink' x='0px' y='0px' viewBox='0 0 10 10' style='enable-background:new 0 0 10 10;' xml:space='preserve'%3E%3Cstyle type='text/css'%3E .st0%7Bfill:%23404040;%7D%0A%3C/style%3E%3Cpolygon class='st0' points='2,2 0,4 5,9 10,4 8,2 5,5 '/%3E%3C/svg%3E%0A");
  background-repeat: no-repeat, no-repeat;
  background-position: left 0.7em top 50%, right 0.7em top 50%;
  background-size: 1em 100%, 0.57em 100%;
  padding-left: 35px;
  border: solid 1px $border-color;
  pointer-events: all;
  margin-left: 20px;
  &:focus {
    background-image: url("data:image/svg+xml,%3C%3Fxml version='1.0' encoding='utf-8'%3F%3E%3C!-- Generator: Adobe Illustrator 24.0.1, SVG Export Plug-In . SVG Version: 6.00 Build 0) --%3E%3Csvg version='1.1' id='Layer_1' xmlns='http://www.w3.org/2000/svg' xmlns:xlink='http://www.w3.org/1999/xlink' x='0px' y='0px' viewBox='0 0 16 16' style='enable-background:new 0 0 16 16;' xml:space='preserve'%3E%3Cstyle type='text/css'%3E .st0%7Bfill:rgba(159, 159, 159, 0.5);%7D .st1%7Bfill:rgba(159, 159, 159, 0.5);%7D%0A%3C/style%3E%3Cg%3E%3Cpath class='st1' d='M8,2.03C3.63,2.03,0,8,0,8s3.63,6.03,8,6.03s8-6,8-6S12.37,2.03,8,2.03z M8,12c-2.58,0-5.13-2.81-6-4 c1.07-1.47,3.61-4,6-4c2.58,0,5.13,2.81,6,4C12.93,9.47,10.39,12,8,12z'/%3E%3Ccircle class='st1' cx='8' cy='8' r='3'/%3E%3C/g%3E%3C/svg%3E"),
      url("data:image/svg+xml,%3C%3Fxml version='1.0' encoding='utf-8'%3F%3E%3C!-- Generator: Adobe Illustrator 21.1.0, SVG Export Plug-In . SVG Version: 6.00 Build 0) --%3E%3Csvg version='1.1' id='Layer_1' xmlns='http://www.w3.org/2000/svg' xmlns:xlink='http://www.w3.org/1999/xlink' x='0px' y='0px' viewBox='0 0 10 10' style='enable-background:new 0 0 10 10;' xml:space='preserve'%3E%3Cstyle type='text/css'%3E .st0%7Bfill:%231AB394;%7D%0A%3C/style%3E%3Cpolygon class='st0' points='2,2 0,4 5,9 10,4 8,2 5,5 '/%3E%3C/svg%3E ") !important;
  }
}

.sa-table__entries {
  float: right;
  white-space: nowrap;
}

.sa-table__entries-label {
  font-weight: bold;
}

.sa-table__entries-label--right {
  margin-right: 5px;
}

.sa-table__entries-label--left {
  margin-left: 5px;
}

.sa-table__detail-row {
  background-color: $detail-color;
}

.sa-table__detail-table {
  background-color: $detail-color;
}

.sa-table input.sa-table__row-extension {
  height: auto;
}

.sa-table__row-extension + .sa-table__row-extension {
  margin-left: 10px;
  vertical-align: middle;
}

.sa-table__header-extensions {
  float: left;
  margin-left: 10px;
  width: calc(100% - 540px);
}

.sa-table__header-extension + .sa-table__header-extension {
  margin-left: 10px;
}

.sa-commercial {
  font-size: 12px;
  padding: 1em 0 0.5em 2em;
  svg {
    width: 20px;
    height: 20px;
    fill: $main-color;
  }
}
.sa-commercial__text {
  font-weight: bold;
  text-decoration: none;
  color: $main-color;
}
.sa-commercial__product {
  padding-left: 0.5em;
}
