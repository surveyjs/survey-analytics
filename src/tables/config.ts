import { Question, SurveyModel } from "survey-core";
import { ITableOptions, Table } from "./table";

export enum QuestionLocation {
  Column,
  Row,
}
export enum ColumnDataType {
  Text,
  FileLink,
  Image,
}

export interface ICellData {
  question: Question;
  displayValue: any;
}

export interface IColumnData {
  name: string;
  displayName: string;
  dataType: ColumnDataType;
  isVisible: boolean;
  isPublic: boolean;
  location: QuestionLocation;
  width?: string | number;
  isComment?: boolean;
}
export interface IColumn extends IColumnData {
  visibleIndex?: number;
  fromJSON(json: any): void;
  getCellData(table: Table, data: any): ICellData;
}

export interface ITableState {
  locale?: string;
  elements?: IColumnData[];
  pageSize?: number;
}

export interface IPermission {
  name: string;
  isPublic: boolean;
}
