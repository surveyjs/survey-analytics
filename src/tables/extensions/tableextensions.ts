import { Table } from "../table";

export interface ITableExtension {
  location: string;
  name: string;
  visibleIndex: number;
  render: (table: Table, opt: any) => HTMLElement;
  destroy?: () => void;
}

export class TableExtensions {
  constructor(private table: Table) {}
  private static extensions: { [location: string]: Array<ITableExtension> } = {};
  private renderedExtensions: Array<ITableExtension> = [];

  public render(targetNode: HTMLElement, location: string, options?: any) {
    var extensions = TableExtensions.extensions[location];
    if (!!extensions) {
      extensions = this.sortExtensions(extensions);
      extensions.forEach((extension) => {
        if (!!extension.render && this.table.allowExtension(extension)) {
          var action = extension.render(this.table, options);
          if (!!action) {
            targetNode.appendChild(action);
            this.renderedExtensions.push(extension);
          }
        }
      });
    }
  }

  public destroy() {
    this.renderedExtensions.forEach((extension) => {
      if (!!extension.destroy) extension.destroy();
    });
    this.renderedExtensions = [];
  }

  public static registerExtension(extension: ITableExtension) {
    if (!this.extensions[extension.location])
      this.extensions[extension.location] = [];
    this.extensions[extension.location].push(extension);
  }

  private static removeExtension(extension: ITableExtension) {
    if(!extension) {
      return;
    }
    const extensions = TableExtensions.extensions[extension.location];
    const index = extensions.indexOf(extension);
    if(index >= 0) {
      extensions.splice(index, 1);
    }
  }

  public static unregisterExtension(
    location: string,
    actionName: string
  ) {
    if(!actionName) {
      return;
    }
    if(!!location) {
      const extension = TableExtensions.findExtension(location, actionName);
      TableExtensions.removeExtension(extension);
    } else {
      Object.keys(this.extensions).forEach((location:string) => TableExtensions.unregisterExtension(location, actionName));
    }
  }

  public static findExtension(
    location: string,
    actionName: string
  ): ITableExtension {
    if (!this.extensions[location]) return null;
    var extension = this.extensions[location].filter(function (
      extension: ITableExtension
    ) {
      return extension.name == actionName;
    })[0];
    return extension || null;
  }

  private sortExtensions(extensions: Array<ITableExtension>) {
    if (!Array.isArray(extensions)) return;
    return []
      .concat(extensions.filter((extension) => extension.visibleIndex >= 0))
      .sort((firstExtension, secondExtension) => {
        return firstExtension.visibleIndex - secondExtension.visibleIndex;
      });
  }
}
