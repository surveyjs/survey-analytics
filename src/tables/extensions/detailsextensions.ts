import { TableRow } from "../table";
import { Table } from "../table";
import { DocumentHelper } from "../../utils";
import { ColumnDataType, QuestionLocation } from "../config";
import { TableExtensions } from "./tableextensions";
import { localization } from "../../localizationManager";

export class Details {
  constructor(
    protected table: Table,
    private row: TableRow,
    protected targetNode: HTMLElement
  ) {
    this.detailsTable = DocumentHelper.createElement(
      "table",
      "sa-table__detail-table"
    );
    this.table.onColumnsLocationChanged.add(() => {
      this.close();
    });
  }
  private detailsTable: HTMLElement;
  protected location = "details";

  public open(): void {
    this.detailsTable.innerHTML = "";
    var rows: HTMLElement[] = [];
    this.table.columns
      .filter((column) => column.location === QuestionLocation.Row && column)
      .forEach((column) => {
        var row = DocumentHelper.createElement("tr", "sa-table__detail");
        var td1 = DocumentHelper.createElement("td");
        td1.appendChild(document.createTextNode(column.displayName));
        var td2 = DocumentHelper.createElement("td");
        td2.textContent = this.row.getRowData()[column.name];
        if(column.dataType === ColumnDataType.Image) {
          td2.innerHTML = "<image src='" + td2.textContent + "'/>";
        }
        var td3 = DocumentHelper.createElement("td");
        td3.appendChild(this.createShowAsColumnButton(column.name));
        row.appendChild(td1);
        row.appendChild(td2);
        row.appendChild(td3);
        rows.push(row);
      });
    var row = DocumentHelper.createElement("tr", "sa-table__detail");
    var td = DocumentHelper.createElement("td", "", { colSpan: 3 });
    var extensions = new TableExtensions(this.table);
    extensions.render(td, "details", { row: this.row });
    if (td.children.length != 0) {
      row.appendChild(td);
      rows.push(row);
    }

    rows.forEach((row) => {
      this.detailsTable.appendChild(row);
    });
    this.targetNode.appendChild(this.detailsTable);
  }

  protected createShowAsColumnButton = (columnName: string): HTMLElement => {
    const button = DocumentHelper.createElement(
      "button",
      "sa-table__btn sa-table__btn--gray"
    );
    button.appendChild(document.createTextNode(localization.getString("showAsColumn")));
    button.onclick = (e) => {
      e.stopPropagation();
      this.table.setColumnLocation(columnName, QuestionLocation.Column);
    };

    return button;
  };

  public close() {
    if (!!this.detailsTable.parentNode) {
      this.detailsTable.parentNode.removeChild(this.detailsTable);
    }
  }
}
