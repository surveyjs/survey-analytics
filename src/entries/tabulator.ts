export * from "../localizationManager";

//localization
import "../analytics-localization/farsi";
import "../analytics-localization/french";
import "../analytics-localization/german";
import "../analytics-localization/norwegian";
import "../analytics-localization/portuguese";
import "../analytics-localization/russian";
import "../analytics-localization/dutch";
import "../analytics-localization/spanish";
import "../analytics-localization/italian";
import "../analytics-localization/arabic";
import "../analytics-localization/japanese";

//extensions
import "../tables/extensions/rowextensions";
import "../tables/extensions/headerextensions";
import "../tables/extensions/columnextensions";
import "../tables/extensions/detailsextensions";

export * from "../tables/table";
export * from "../tables/tabulator";
export { TableExtensions } from "../tables/extensions/tableextensions";
export { DocumentHelper } from "../utils/index";
