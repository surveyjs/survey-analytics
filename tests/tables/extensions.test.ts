import { SurveyModel } from "survey-core";
import { Tabulator, TabulatorRow } from "../../src/tables/tabulator";
import { TableExtensions } from "../../src/tables/extensions/tableextensions";
import { Details } from "../../src/tables/extensions/detailsextensions";
import { TableRow } from "../../src/tables/table";
import { DocumentHelper } from "../../src/utils";
import { QuestionLocation } from "../../src/tables/config";

export * from "../../src/tables/extensions/headerextensions";
export * from "../../src/tables/extensions/footerextensions";
export * from "../../src/analytics-localization/german";

const json = {
  questions: [
    {
      type: "radiogroup",
      name: "car",
      title: "What car are you driving?",
      isRequired: true,
      colCount: 4,
      choices: ["None", "Ford", "Vauxhall"],
    },
    {
      type: "file",
      name: "photo",
    },
  ],
};
const survey = new SurveyModel(json);
const tabulator = new Tabulator(survey, [], null);
(<any>TableExtensions).extensions["row"] = [];

test("registerExtension method", () => {
  var extension: any = {
    name: "test",
    location: "row",
    visibleIndex: 0,
    render: null,
  };
  TableExtensions.registerExtension(extension);
  expect((<any>TableExtensions).extensions["row"][0]).toEqual(extension);
  (<any>TableExtensions).extensions["row"].splice(0, 1);
});

test("check extensions order with same index", () => {
  TableExtensions.registerExtension(<any>{
    name: "prev",
    location: "row",
    visibleIndex: 0,
    render: null,
  });
  TableExtensions.registerExtension(<any>{
    name: "next",
    location: "row",
    visibleIndex: 0,
    render: null,
  });
  var tableExtensions = new TableExtensions(tabulator);
  expect(
    (<any>tableExtensions)
      .sortExtensions((<any>TableExtensions).extensions["row"])
      .map((extension: any) => {
        return extension.name;
      })
  ).toEqual(["prev", "next"]);
  (<any>TableExtensions).extensions["row"].splice(0, 2);
});

test("check extensions order index -1", () => {
  TableExtensions.registerExtension({
    name: "visible",
    location: "row",
    visibleIndex: 0,
    render: null,
  });

  TableExtensions.registerExtension({
    name: "invisible",
    location: "row",
    visibleIndex: -1,
    render: null,
  });
  var tableExtensions = new TableExtensions(tabulator);
  expect(
    (<any>tableExtensions)
      .sortExtensions((<any>TableExtensions).extensions["row"])
      .map((extension: any) => {
        return extension.name;
      })
  ).toEqual(["visible"]);
  (<any>TableExtensions).extensions["row"].splice(0, 2);
});

test("check findExtension method", () => {
  var extension: any = {
    name: "test",
    location: "row",
    visibleIndex: 0,
    render: null,
  };
  TableExtensions.registerExtension(extension);
  expect(TableExtensions.findExtension("row", "test")).toEqual(extension);
  (<any>TableExtensions).extensions["row"].splice(0, 1);
});

test("check rendering extension with null render", () => {
  var extension: any = {
    name: "test",
    location: "header",
    visibleIndex: 0,
    render: null,
  };
  TableExtensions.registerExtension(extension);
  var tableExtensions = new TableExtensions(tabulator);
  expect(() => {
    tableExtensions.render(
      document.createElement("extensionsContainer"),
      "header"
    );
  }).not.toThrow(Error);
  (<any>TableExtensions).extensions["header"].splice(0, 1);
});

test("check rendering extension with render's null return value", () => {
  var extension: any = {
    name: "test",
    location: "header",
    visibleIndex: 0,
    render: (): any => {
      return null;
    },
  };
  TableExtensions.registerExtension(extension);
  var tableExtensions = new TableExtensions(tabulator);
  expect(() => {
    try {
      tableExtensions.render(
        document.createElement("extensionsContainer"),
        "header"
      );
    } catch {
      throw new Error("tried to append null child");
    }
  }).not.toThrow(Error);
  (<any>TableExtensions).extensions["header"].splice(0, 1);
});

test("render image in details", () => {
  const survey = new SurveyModel({ elements: [{ type: "signaturepad", name: "q1" }] });
  const tabulator = new Tabulator(survey, [], null);
  tabulator.columns[0].location = QuestionLocation.Row;
  const extensionsContainer = DocumentHelper.createElement("div");
  const detailsContainer = DocumentHelper.createElement("div");
  const tableRow = new TabulatorRow(tabulator, extensionsContainer, detailsContainer, { getData: () => ({ q1: "signature" }) });
  const detailsTarget = DocumentHelper.createElement("div");
  const details = new Details(tabulator, tableRow, detailsTarget);
  details.open();
  expect(detailsTarget).toMatchSnapshot();
});

test("locale selector uses titles", () => {
  const survey = new SurveyModel({ elements: [{ type: "signaturepad", name: "q1", title: { "default": "EnTitle", "de": "DeTitle" } }] });
  const tabulator = new Tabulator(survey, []);
  const changeLocaleHeaderExtension = TableExtensions.findExtension("header", "changelocale");
  expect(changeLocaleHeaderExtension).toBeDefined();
  const renderResult = changeLocaleHeaderExtension.render(tabulator, undefined) as HTMLSelectElement;
  expect(renderResult.options.length).toBe(3);
  expect(renderResult.options[0].text).toBe("Change Locale");
  expect(renderResult.options[1].text).toBe("English");
  expect(renderResult.options[2].text).toBe("Deutsch");
});

test("changelocale respects disableLocaleSwitch", () => {
  const survey = new SurveyModel({ elements: [{ type: "text", name: "q1", title: {
    default: "Qd",
    en: "Qe",
    fr: "Qf",
  } }] });
  const tabulator = new Tabulator(survey, [], { disableLocaleSwitch: true } as any);
  const changeLocaleExtension = TableExtensions.findExtension("header", "changelocale");

  expect(changeLocaleExtension.render(tabulator, null)).toBe(null);

  tabulator.options.disableLocaleSwitch = false;
  expect(changeLocaleExtension.render(tabulator, null)).toBeDefined();
  expect(changeLocaleExtension.render(tabulator, null).tagName).toBe("SELECT");
});
