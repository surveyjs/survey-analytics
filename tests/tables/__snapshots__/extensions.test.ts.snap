// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`render image in details 1`] = `
<div
  class=""
>
  <table
    class="sa-table__detail-table"
  >
    <tr
      class="sa-table__detail"
    >
      <td
        class=""
      >
        <button
          class="sa-table__svg-button"
          title="Show as Column"
        >
          <svg>
            <use
              href="#sa-svg-movetovertical-24x24"
            />
          </svg>
        </button>
      </td>
      <td
        class=""
      >
        q1
      </td>
      <td
        class=""
      >
        <img
          src="signature"
        />
      </td>
    </tr>
  </table>
</div>
`;
